spoc/contacts.py

```diff
@@ -1,84 +1,289 @@
+                output.update(self.data[f"meta_data_{i+1}"].unique())
         return output
 
-    def expand(self, fragments: pd.DataFrame) -> pd.DataFrame:
-        """expand contacts n-ways"""
-        # check input
-        AnnotatedFragmentSchema.validate(fragments)
-        # expand fragments
-        result = []
-        keep_segments = fragments.query("pass_filter == True")
-        for (read_name, read_df) in keep_segments.groupby("read_name", as_index=False):
-            if len(read_df) < self._number_fragments:
-                continue
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, contact_frame):
+        self._data = self._schema.validate(contact_frame)
 
-            rows = list(
-                read_df.sort_values(["read_start"], ascending=True)
-                .assign(pos_on_read=lambda x: np.arange(len(x)))
-                .itertuples()
-            )
 
-            read_length = rows[0].read_length
-            for alignments in combinations(rows, self._number_fragments):
-                contact = {"read_name": read_name, "read_length": read_length}
-                # add reads
-                for index, align in enumerate(alignments, start=1):
-                    contact.update(self._add_suffix(align, index))
-                result.append(contact)
-        return pd.DataFrame(result)
+    def __repr__(self) -> str:
+        return f"<Contacts | order: {self.number_fragments} | contains metadata: {self.contains_meta_data}>"
 
 
 class ContactManipulator:
     """Responsible for performing operations on
     contact data such as merging, splitting and subsetting."""
 
-    def __init__(self, number_fragments: int, use_dask: bool = False) -> None:
-        self._number_fragments = number_fragments
-        self._schema = HigherOrderContactSchema(number_fragments=number_fragments)
-        self._use_dask = use_dask
-
-    def _merge_contacts_pandas(self, merge_list: List[pd.DataFrame]) -> pd.DataFrame:
-        # validate schema
-        for data_frame in merge_list:
-            self._schema.validate(data_frame)
-        return pd.concat(merge_list)
-
-    def _merge_contacts_dask(self, merge_list: List[dd.DataFrame]) -> dd.DataFrame:
-        # validate schema
-        merge_with_check = []
-        for data_frame in merge_list:
-            merge_with_check.append(
-                self._schema.validate(data_frame)
-            )  # needed to add check to task graph
-        return dd.concat(merge_with_check)
-
-    def merge_contacts(
-        self, merge_list: List[Union[pd.DataFrame, dd.DataFrame]]
-    ) -> Union[pd.DataFrame, dd.DataFrame]:
+    def merge_contacts(self, merge_list: List[Contacts]) -> Contacts:
         """Merge contacts"""
-        if not self._use_dask:
-            return self._merge_contacts_pandas(merge_list)
-        return self._merge_contacts_dask(merge_list)
+        # validate that merge is possible
+        assert (
```

Comment:
Should we change `assert` to `raise ValueError` (or smth else) in the future? It seems like these lines validate user input, not the code itself.

Reply:
Agreed, addressed in 4804445c56df1bc2a05ea37cd56fa22a4f5beef3